pkg/services/sqlstore/migrations/object_store_mig.go

```diff
@@ -0,0 +1,157 @@
+			{Name: "labels", Type: migrator.DB_Text, Nullable: true}, // JSON object
+			{Name: "fields", Type: migrator.DB_Text, Nullable: true}, // JSON object
+			{Name: "errors", Type: migrator.DB_Text, Nullable: true}, // JSON object
+		},
+		PrimaryKeys: []string{"key"},
+		Indices: []*migrator.Index{
+			{Cols: []string{"parent_folder_key"}}, // list in folder
+			{Cols: []string{"kind"}},              // filter by type
+		},
+	}
+
+	objectLabelsTable := migrator.Table{
+		Name: "object_labels",
+		Columns: []*migrator.Column{
+			{Name: "key", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false},
+			{Name: "label", Type: migrator.DB_NVarchar, Length: 191, Nullable: false},
+			{Name: "value", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false},
+		},
+		Indices: []*migrator.Index{
+			{Cols: []string{"key", "label"}, Type: migrator.UniqueIndex},
+		},
+	}
+
+	objectReferenceTable := migrator.Table{
+		Name: "object_ref",
+		Columns: []*migrator.Column{
+			// Source:
+			{Name: "key", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false},
+
+			// Address (defined in the body, not resolved, may be invalid and change)
+			{Name: "kind", Type: migrator.DB_NVarchar, Length: 255, Nullable: false},
+			{Name: "type", Type: migrator.DB_NVarchar, Length: 255, Nullable: true},
+			{Name: "uid", Type: migrator.DB_NVarchar, Length: 1024, Nullable: true},
+
+			// Runtime calcs (will depend on the system state)
+			{Name: "resolved_ok", Type: migrator.DB_Bool, Nullable: false},
+			{Name: "resolved_to", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false},
+			{Name: "resolved_warning", Type: migrator.DB_NVarchar, Length: 255, Nullable: false},
+			{Name: "resolved_time", Type: migrator.DB_DateTime, Nullable: false}, // resolution cache timestamp
+		},
+		Indices: []*migrator.Index{
+			{Cols: []string{"key"}, Type: migrator.IndexType},
+			{Cols: []string{"kind"}, Type: migrator.IndexType},
+			{Cols: []string{"resolved_to"}, Type: migrator.IndexType},
+		},
+	}
+
+	objectHistoryTable := migrator.Table{
+		Name: "object_history",
+		Columns: []*migrator.Column{
+			{Name: "key", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false},
+			{Name: "version", Type: migrator.DB_NVarchar, Length: 128, Nullable: false},
+
+			// Raw bytes
+			{Name: "body", Type: migrator.DB_Blob, Nullable: false},
+			{Name: "size", Type: migrator.DB_BigInt, Nullable: false},
+			{Name: "etag", Type: migrator.DB_NVarchar, Length: 32, Nullable: false}, // md5(body)
+
+			// Who changed what when
+			{Name: "updated", Type: migrator.DB_DateTime, Nullable: false},
+			{Name: "updated_by", Type: migrator.DB_Int, Nullable: false},
+
+			// Commit message
+			{Name: "message", Type: migrator.DB_Text, Nullable: false}, // defaults to empty string
+		},
+		Indices: []*migrator.Index{
+			{Cols: []string{"key", "version"}, Type: migrator.UniqueIndex},
+		},
+	}
+
+	// Define access based on prefix rules (eg... .htaccess (- ‿◦ ))
```

Comment:
this is still exploratory, and will take some work to know if it makes sense to continue.  For sure authn should be managed by a separate service, but I am not yet convinced authz should be -- this will matter more if we choose to use path based authorization.  The nice part of this is that it would be more self contained and have all the same history and audit controls.

That said -- I am far from convinced this is the right path forward, but it is much simpler and easier to embed/control than our current approach